matrix-sdk/src/main/java/org/matrix/androidsdk/crypto/MXMegolmExportEncryption.java

```diff
@@ -0,0 +1,362 @@
+        Mac mac = Mac.getInstance("HmacSHA256");
+        mac.init(macKey);
+        byte[] digest = mac.doFinal(toVerify);
+
+        if (!Arrays.equals(hmac, digest)) {
+            Log.e(LOG_TAG, "## decryptMegolmKeyFile() : Authentication check failed: incorrect password?");
+            throw new Exception("Authentication check failed: incorrect password?");
+        }
+
+        Cipher decryptCipher = Cipher.getInstance("AES/CTR/NoPadding");
+
+        SecretKeySpec secretKeySpec = new SecretKeySpec(getAesKey(deriveKey), "AES");
+        IvParameterSpec ivParameterSpec = new IvParameterSpec(iv);
+        decryptCipher.init(Cipher.DECRYPT_MODE, secretKeySpec, ivParameterSpec);
+
+        ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+        outStream.write(decryptCipher.update(ciphertext));
+        outStream.write(decryptCipher.doFinal());
+
+        String decodedString = new String(outStream.toByteArray(), "UTF-8");
+        outStream.close();
+
+        return decodedString;
+    }
+
+    /**
+     * Encrypt a string into the megolm export format.
+     *
+     * @param data     the data to encrypt.
+     * @param password the password
+     * @return the encrypted data
+     * @throws Exception the failure reason
+     */
+    public static byte[] encryptMegolmKeyFile(String data, String password) throws Exception {
+        return encryptMegolmKeyFile(data, password, 100000);
+    }
+
+    /**
+     * Encrypt a string into the megolm export format.
+     *
+     * @param data       the data to encrypt.
+     * @param password   the password
+     * @param kdf_rounds the iteration count
+     * @return the encrypted data
+     * @throws Exception the failure reason
+     */
+    public static byte[] encryptMegolmKeyFile(String data, String password, int kdf_rounds) throws Exception {
+        if (TextUtils.isEmpty(password)) {
+            throw new Exception("Empty password is not supported");
+        }
+
+        SecureRandom secureRandom = new SecureRandom();
+
+        byte[] salt = new byte[16];
+        secureRandom.nextBytes(salt);
+
+        byte[] iv = new byte[16];
+        secureRandom.nextBytes(iv);
+
+        // clear bit 63 of the salt to stop us hitting the 64-bit counter boundary
+        // (which would mean we wouldn't be able to decrypt on Android). The loss
+        // of a single bit of salt is a price we have to pay.
+        iv[9] &= 0x7f;
+
+        byte[] deriveKey = deriveKeys(salt, kdf_rounds, password);
+
+        Cipher decryptCipher = Cipher.getInstance("AES/CTR/NoPadding");
+
+        SecretKeySpec secretKeySpec = new SecretKeySpec(getAesKey(deriveKey), "AES");
+        IvParameterSpec ivParameterSpec = new IvParameterSpec(iv);
+        decryptCipher.init(Cipher.ENCRYPT_MODE, secretKeySpec, ivParameterSpec);
+
+        ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+        outStream.write(decryptCipher.update(data.getBytes("UTF-8")));
```

Comment:
To save RAM, it may worth to cipher the input data by packets rather than the whole input buffer in a single pass.